TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,273 @@
+﻿using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Security;
+using Org.BouncyCastle.Utilities;
+using System;
+
+
+namespace TumbleBitSetup
+{
+    public class PoupardStern
+    {
+        /// <summary>
+        /// Proving Algorithm specified in page 12 (3.2.1) of the setup
+        /// </summary> 
+        /// <param name="p">P in the secret key</param>
+        /// <param name="q">Q in the secret key</param>
+        /// <param name="e">Public Exponent in the public key</param>
+        /// <param name="ps">The public string from the setup</param>
+        /// <param name="k">Security parameter as specified in the setup.</param>
+        /// <returns>List of x values and y</returns>
+        public static Tuple<BigInteger[], BigInteger> Proving(BigInteger p, BigInteger q, BigInteger e, int keyLength, byte[] psBytes, int k = 128)
+        {
+            k = Utils.GetByteLength(k) * 8;
+            BigInteger y;
+            BigInteger Two = BigInteger.Two;
+            // 2^{|N| - 1}
+            BigInteger lowerLimit = Two.Pow(keyLength - 1);
+            // 2^{|N|}
+            BigInteger upperLimit = Two.Pow(keyLength);
+
+            // Generate a keyPair from p, q and e
+            var keyPair = new RsaKey(p, q, e);
+            var pubKey = new RsaPubKey(keyPair._pubKey);
+            var secKey = keyPair._privKey;
+
+            BigInteger Modulus = pubKey._pubKey.Modulus;
+
+            // Check if N <= 2^{|N|-1}
```

Comment:
OK, I can live with this "CompareTo" thing as long as we don't also use "not" when we "CompareTo"